blaze/server/server.py

```diff
@@ -393,38 +493,101 @@ def from_tree(expr, namespace=None):
-    except Exception as e:
-        # 500: Internal Server Error
+    (allow_profiler,
+     default_profiler_output,
+     profile_by_default) = _get_profiler_info()
+    requested_profiler_output = payload.get(
+        'profiler_output',
+        default_profiler_output,
+    )
+    profile = payload.get('profile')
+    profiling = (
+        allow_profiler and
+        (profile or (profile_by_default and requested_profiler_output))
+    )
+    if profile and not allow_profiler:
         return (
-            "Computation failed with message:\n%s: %s" % (type(e).__name__, e),
-            500,
+            'profiling is disabled on this server',
+            403,
         )
 
-    return serial.dumps({
-        'datashape': pprint(expr.dshape, width=0),
-        'data': result,
-        'names': expr.fields
-    })
+    with ExitStack() as stack:
+        if profiling:
+            from cProfile import Profile
+
+            if (default_profiler_output == ':response' and
+                requested_profiler_output != ':response'):
+                # writing to the local filesystem is disabled
+                return (
+                    "local filepaths are disabled on this server, only"
+                    " ':response' is allowed for the 'profiler_output' field",
+                    403,
+                )
+
+            profiler_output = requested_profiler_output
+            profiler = Profile()
+            profiler.enable()
+            # ensure that we stop profiling in the case of an exception
+            stack.callback(profiler.disable)
+
+        ns = payload.get('namespace', dict())
+        compute_kwargs = payload.get('compute_kwargs') or {}
+        odo_kwargs = payload.get('odo_kwargs') or {}
+        dataset = _get_data()
+        ns[':leaf'] = symbol('leaf', discover(dataset))
+
+        expr = from_tree(payload['expr'], namespace=ns)
+        assert len(expr._leaves()) == 1
+        leaf = expr._leaves()[0]
+
+        try:
+            result = serial.materialize(
+                compute(expr, {leaf: dataset}, **compute_kwargs),
+                expr.dshape,
+                odo_kwargs,
+            )
+        except NotImplementedError as e:
+            # 501: Not Implemented
+            return ("Computation not supported:\n%s" % e, 501)
+        except Exception as e:
+            # 500: Internal Server Error
+            return (
+                "Computation failed with message:\n%s: %s" % (
+                    type(e).__name__,
+                    e
+                ),
+                500,
+            )
+
+        response = {
+            'datashape': pprint(expr.dshape, width=0),
+            'data': serial.data_dumps(result),
+            'names': expr.fields
+        }
+        if profiling:
+            import marshal
+            from pstats import Stats
+
+            # remove our callback because we are going to exit now
+            stack.pop_all()
```

Comment:
Is it worth making this more robust to people adding other ctx managers to the stack?  Maybe exit the stack unconditionally after producing the `response`?